include/lock/optimistic_lock.hpp

```diff
@@ -209,23 +209,29 @@ class OptimisticLock
   /**
    * @brief Downgrade an X lock to an SIX lock.
    *
+   * @retval version value after downgrade from an X lock to an SIX lock.
+   *
    * NOTE: if a thread that does not have an exclusive lock calls this function, it will
    * corrupt an internal lock status.
    */
-  void
-  DowngradeToSIX()
+  auto
+  DowngradeToSIX()  //
+      -> uint64_t
   {
-    lock_.fetch_add(kXLock + kSIXLock, std::memory_order_release);
+    return (lock_.fetch_add(kXLock + kSIXLock, std::memory_order_release) + kXLock + kSIXLock)
+           & kAllBitsMask;
```

Comment:
見通しが悪くなっちゃうので，一旦局所変数に入れたほうがよいですね．

```suggestion
    const auto old_ver = lock_.fetch_add(kXLock + kSIXLock, std::memory_order_release);
    return (old_ver + kXLock + kSIXLock) & kAllBitsMask;
```